x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotRepository.java

```diff
@@ -160,16 +164,17 @@ public Repository create(RepositoryMetaData metaData, Function<String, Factory>
         return (indexSettings, shardPath) -> {
             final RepositoriesService repositories = repositoriesService.get();
             assert repositories != null;
-
-            final Repository repository = repositories.repository(SNAPSHOT_REPOSITORY_SETTING.get(indexSettings.getSettings()));
-            if (repository instanceof SearchableSnapshotRepository == false) {
-                throw new IllegalArgumentException("Repository [" + repository + "] is not searchable");
-            }
-
             final CacheService cache = cacheService.get();
             assert cache != null;
 
-            return ((SearchableSnapshotRepository) repository).makeDirectory(indexSettings, shardPath, cache, currentTimeNanosSupplier);
+            final Repository repository = repositories.repository(SNAPSHOT_REPOSITORY_SETTING.get(indexSettings.getSettings()));
+            // TODO: is this okay?
+//            if (repository instanceof SearchableSnapshotRepository == false) {
+//                throw new IllegalArgumentException("Repository [" + repository + "] is not searchable");
+//            }
+
+            SearchableSnapshotRepository searchableRepo = new SearchableSnapshotRepository(repository);
```

Comment:
This is "disrupting" the usual lifecycle of `Repository`. On the other hand, `SearchableSnapshotRepository` was useful when the searchable repository was explicitly registered as a delegating repository. I don't think it's needed anymore: the  `DirectoryFactory` logic can be moved back to the plugin class, and instead of passing directly a `BlobContainer` instance to the `SearchableSnapshotDirectory` it could use the `RepositoriesService` to retrieve it on demande. This way we get rid of a class and we take care of reloading the `BlobStoreRepository` instance that could have changed.

Reply:
I would leave that up to you and David. I agree this is not the "right" way to do it since there is no longer a need for a `SearchableSnapshotRepository`, I agree that the directory logic can be moved elsewhere when the `SearchableSnapshotRepository` repository type is removed. For now, I was trying to make as small a change as possible to code outside of the scope of the actual request/response/transport.

Reply:
I'm with Tanguy here, I think it would be good to remove `SearchableSnapshotRepository` in this PR.